extensions/amp-smartlinks/0.1/linkmate.js

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ENDPOINTS} from './constants';
+import {TwoStepsResponse} from './link-rewriter/two-steps-response';
+import {deepEquals} from '../../../src/json';
+
+
+export class Linkmate {
+  /**
+   * @param {?../../../src/service/ampdoc-impl.AmpDoc} ampDoc
+   * @param {!../../../src/service/xhr-impl.Xhr} xhr
+   * @param {?Object} linkmateOptions
+   */
+  constructor(ampDoc, xhr, linkmateOptions) {
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = ampDoc;
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = xhr;
+
+    /** @private {?boolean} */
+    this.requestExclusiveLinks_ = linkmateOptions.exclusiveLinks;
+
+    /** @private {!int} */
+    this.publisherID_ = linkmateOptions.publisherID;
+
+    /** @private {string} */
+    this.linkAttribute_ = linkmateOptions.linkAttribute;
+
+    /** @private {!Document|!ShadowRoot} */
+    this.rootNode_ = this.ampDoc_.getRootNode();
+
+    /** @private {!Array<!HTMLElement>} */
+    this.anchorList_ = null;
+
+    /** @private {!Array<JsonObject>}*/
+    this.linkmateResponse_ = null;
+  }
+
+  /**
+   * Callback used by LinkRewriter. Whenever there is a change in the anchors
+   * on the page we want make a new API call.
+   * @param {!Array<!HTMLElement>} anchorList
+   * @return {!./link-rewriter/two-steps-response.TwoStepsResponse}
+   * @public
+   */
+  runLinkmate(anchorList) {
```

Comment:
To me, it looks like if 1 href is added or removed, all hrefs on the page will be sent to the API again. Is that the intention?

Reply:
Yeah that was the intention. I can add more logic to limit the API call but iirc from early design decisions we decided to be naive and just resend the full payload.

Reply:
OK